# Question 2: International Coding Competition

Name this Jupyter Notebook as: <br>
`TASK2_<your name>_<class>_<index number>.ipynb`

For each of the sub-tasks, add a comment statement at the beginning of the code using the hash symbol '#' to indicate the sub-task the program code belongs to.

## Task 2.1
Write an insertion sort function that takes three parameters: `data_list`,`index`,`ascending` and returns a sorted `data_list`.

* `data_list` is a list of tuples.
* `index` is an integer to specify an index within a tuple.
* `ascending` is a Boolean flag.

The function will sort `data_list` base on the values at the specified `index` within the tuples. The `ascending` parameter determins the order of the sort: `True` for ascending order, `False` for descending order.

The function specification is:

```python
FUNCTION insertion_sort(data_list: List[Tuple], index: Integer, ascending:Boolean) RETURN List[Tuple]
```
| Note: The syntax that follows the positional argument is just type hinting.

<div style="text-align: right">[6]</div>

> Example of `insertion_sort` function call:
>```python
> data_list = [(1,'Alice', 85), (2, 'Bob',90), (3,'Charlie',80)]
> sorted_list = insertion_sort(data_list, 2, ascending=False)
> print(sorted_list)
>
> # Output: [(2,'Bob',90), (1,'Alice',85), (3,'Charlie',80)]
>```



In [6]:
# Task 2.1

def insertion_sort(data_list,index,ascending=True):
    length = len(data_list)
    for i in range(1,length):
        for j in range(i,0,-1):
            if data_list[j][index] < data_list[j-1][index] and ascending:
                data_list[j],data_list[j-1] = data_list[j-1],data_list[j]
            elif data_list[j][index] > data_list[j-1][index] and not ascending:
                data_list[j],data_list[j-1] = data_list[j-1],data_list[j]
            else:
                break
    return data_list
data_list = [(1,'Alice', 85), (2, 'Bob',90), (3,'Charlie',80)]
insertion_sort(data_list,2,ascending=False)                  

[(2, 'Bob', 90), (1, 'Alice', 85), (3, 'Charlie', 80)]

## Task 2.2
At an international coding competition, the results of the participants were recorded in a text file `CODING_COMPETITION.txt`.

Each record int he file has the following format, with each field separated by a comma:
`<participant_id>,<participant_name>,<country>,<score>`

Write a function `task2_2` that takes a `filename` as parameter, reads the comma-separated data from the file and returns the data as a list of tuples:

> Example: After calling function `task2_2`, these sample records in the file: <br>
>
> `101,Jong Tan,Singapore,210` <br>
> `102,Mei Ling,Singapore,190` <br>
>
> will be returned as <br>
> `[('101','Jong Tan','Singapore',210),('102','Mei Ling','Singapore',190)]`

Test your function by calling `task2_2("CODING_COMPETITION.txt")`
<div style="text-align: right">[5]<div>

In [13]:
# Task 2.2
import csv

def task2_2(filename):
    file_path = f"./Resources/TASK2/{filename}"
    with open(file_path) as file:
        reader = csv.reader(file)
        data = list(reader)

    # For loop    
    for i in range(len(data)):
        data[i][-1] = int(data[i][-1])
        data[i] = tuple(data[i])
        
    # List comprehension
    data = [(*d[:-1],int(d[-1])) for d in data]

    return data

task2_2("CODING_COMPETITION.txt")



[('101', 'Jong Tan', 'Singapore', 210),
 ('102', 'Mei Ling', 'Singapore', 190),
 ('103', 'John Smith', 'USA', 230),
 ('104', 'Emily Davis', 'USA', 85),
 ('105', 'Chen Wei', 'China', 220),
 ('106', 'Li Na', 'China', 105),
 ('107', 'Kumar Patel', 'India', 100),
 ('108', 'Arjun Rao', 'India', 95),
 ('109', 'Alice Johnson', 'UK', 240),
 ('110', 'Robert Brown', 'UK', 70),
 ('111', 'Marco Rossi', 'Italy', 225),
 ('112', 'Giulia Ferrari', 'Italy', 60),
 ('113', 'Ahmed Khan', 'Pakistan', 210),
 ('114', 'Ali Ahmed', 'Pakistan', 80),
 ('115', 'Olga Ivanova', 'Russia', 220),
 ('116', 'Dmitry Petrov', 'Russia', 90),
 ('117', 'Paul Walker', 'Australia', 215),
 ('118', 'Emily Brown', 'Australia', 85),
 ('119', 'Julia Schmidt', 'Germany', 235),
 ('120', 'Matthias Mueller', 'Germany', 55),
 ('121', 'Yuki Tanaka', 'Japan', 225),
 ('122', 'Haruka Sato', 'Japan', 95),
 ('123', 'Karen Wilson', 'Canada', 250),
 ('124', 'David Wilson', 'Canada', 70),
 ('125', 'Lucia Martinez', 'Spain', 205),
 ('126', 'Javie

## Task 2.3
Write a function `task2_3` that performs the following:
* Use the list of tuples returned from the function `task2_2`.
* Sort the list by country in alphabetical order, using the `insertion_sort` function
* Within each country, sort the participants by their scores in descending order using the `insertion_sort` function.
* Write the sorted data to a text file `SCORE_BY_COUNTRY.txt` in the format: <br>
`<country>,<participant id>,<participant name>,<score>`.
* Each record should be on a new line

The output file should contain lines of record that is first sorted by country then by score. Sample lines in output file:

<center>

<table border="1">
  <tr style="font-family: 'Courier New', monospace;">
    <td>Australia,249,James Scott, 235<br>
        Australia,267,Liam Taylor, 230<br>
        Australia,139,Daniel Miller, 225<br>
        ...<br>
        Singapore,101,Jong Tan,210<br>
        ...</td>
  </tr>
</table>

</center>

Run your function `task2_3` 
<div style="text-align:right;">[9]</div>

In [ ]:
# Task 2.3
import csv

def task2_3(data):
    lib = {}
    for i in data:
        country = i[2]
        if country not in lib.keys():
            lib[country] = [i]
        else:
            lib[country] = lib[country] + [i]
    countries_by_alpha = insertion_sort(list(lib.keys()),0,ascending = True)
    w = []
    for country in countries_by_alpha:
        w += insertion_sort(lib[country],-1,ascending=False)
    with open("./Resources/TASK2/SCORE_BY_COUNTRY.txt","w",newline="") as file:
        writer = csv.writer(file)
        writer.writerows(w)

data = task2_2("CODING_COMPETITION.txt")
task2_3(data)
    
